docs/design/kata-nydus-design.md

```diff
@@ -0,0 +1,91 @@
+
+The process of creating/starting Kata Containers with virtiofsd,
+
+1. when creating sandbox, the Kata Containers shim will launch virtiofsd before VM starts, so VM can share directories with host.
+2. when creating container, the Kata Containers shim will mount rootfs to /path/to/bundle/rootfs and bind mount rootfs to `sharedir`, so it can be seen in the guest.
+
+The process of creating/starting Kata Containers with nydusd,
+
+![kata-nydus](arch-images/kata-nydus.png)
+
+1. when creating sandbox, the Kata Containers Containerd v2 shim starts `nydusd` daemon and mounts `passthroughfs` filesystem before the VM starts.
+
+```bash
+# start nydusd
+$ sudo /usr/local/bin/nydusd --log-level info  
+    --sock /run/vc/vm/${sandboxID}/vhost-user-fs.sock 
+    --apisock  /run/vc/vm/${sandboxID}/api.sock
+```
+
+```bash
+# mount passthroughfs
+$ curl -v --unix-socket /run/vc/vm/${sandboxID}/api.sock \
+    -X POST "http://localhost/api/v1/mount?mountpoint=/foo1" -H 'accept: */*' \
+    -H "Content-Type: application/json" \
+    -d '{
+            "source":"/path/to/sharedir",
+            "fs_type":"passthrough_fs",
+            "config":""
+    }'
+```
+
+2. when creating normal container, the Kata Containers Containerd v2 shim send request to nydusd to mount `rafs` to guest. The Kata Containers Containerd v2 shim will also bind mount `snapshotdir` which `nydus-snapshotter` assigns to `sharedir`, so in guest, container rootfs=overlay(lowerdir=`rafs`, upperdir=`snapshotdir/fs`, workdir=`snapshotdir/work`)
+
+```bash
+# mount rafs
+# source: the metafile of nydus image
+# config: the config of this image
+$ curl --unix-socket /run/vc/vm/${sandboxID}/api.sock \
+    -X POST "http://localhost/api/v1/mount?mountpoint=/bar1" -H 'accept: */*' \
+    -H "Content-Type: application/json" \
+    -d '{
+            "source":"/path/to/bootstrap",
+            "fs_type":"rafs",
+            "config":"{}",
+    }'
+```
+
+> So the question is `how to transfer the rafs info from nydus-snapshotter to the Kata Containers Containerd v2 shim?`
+
+By default, when creating `OCI` image container, `nydus-snapshotter` will return [struct Mount slice](https://github.com/containerd/containerd/blob/main/mount/mount.go#L21) below to containerd and containerd use them to mount rootfs
+
+```bash
+[
+    {
+        Type: "overlay",
+        Source: "overlay",
+        Options: [lowerdir=lower_A:lower_B,upperdir=upper_A,workdir=work_A],
+    }
+]
+```
+
+Then, we can append rafs info into `Options`, but if do this, containerd will mount failed, as containerd can not identify rafs info. Here, we can refer to [containerd mount helper](https://github.com/containerd/containerd/blob/main/mount/mount_linux.go#L42) and provide a binary called `nydus-overlayfs`. The `Mount` slice which nydus-snapshotter returned becomes
+
+```bash
```

Comment:
JSON?

Reply:
this is pseudo code, not json doc strictly. I try `json`, it will show unexpectedly

Reply:
Ok. I'd remove the `bash` then and just use ` ``` `.

Reply:
done